src/daemons/linuxd/src/linuxd/mod.rs

```diff
@@ -157,26 +159,101 @@ pub struct LinuxDaemon {
 impl LinuxDaemon {
     pub fn init(
         uvm_stream: SocketStream,
-        gateway_conn: Option<SocketStream>,
+        gateway_stream: Option<SocketStream>,
     ) -> Result<Self, Error> {
         if let Err(error) = uvm_stream.set_nonblocking(true) {
             let reason: &str = "failed to set UVM stream to non-blocking mode";
             error!("init(): {reason:?} (error={error:?})");
             return Err(Error::new(ErrorCode::InvalidArgument, reason));
         }
 
+        let gateway_stream_opt = gateway_stream.map(|stream| Arc::new(Mutex::new(stream)));
+
         Ok(Self {
             assembler: Arc::new(Mutex::new(RequestAssembler::default())),
             uvm_stream: Arc::new(Mutex::new(uvm_stream)),
-            gateway_conn: Arc::new(Mutex::new(gateway_conn)),
+            gateway_stream: gateway_stream_opt,
             venv: Arc::new(Mutex::new(VirtualEnviromentDirectory::new())),
         })
     }
 
     pub fn run(&mut self) -> Result<(), Error> {
+        // Start the thread that will poll input from the gateway.
+        // TODO: when one linuxd instance manages more than one input stream we should encapsulate
+        // this logic.
+        let (gw_stdin_tx, gw_stdout_tx) = if let Some(gateway_stream) = self.gateway_stream.clone() {
+            let (gw_stdin_tx, gw_stdin_rx) = mpsc::channel::<(ReadRequest, Sender<Message>)>();
+            let (gw_stdout_tx, gw_stdout_rx) = mpsc::channel::<(WriteRequest, Sender<Message>)>();
+
+            // Make sure that the input stream from the gateway is set to blocking, as otherwise we
+            // would not be able to differentiate between an EOF and a race between the application
+            // code and the gateway.
+            let gw_stdin_stream = gateway_stream.clone();
+            gw_stdin_stream
+                .lock()
+                .unwrap()
+                .set_nonblocking(false)
+                .unwrap();
```

Comment:
We should enforce this at creation time and enforce semantics using typestate: `BlockingSocketStream`. We can take it in a different PR. If so, could you just create an issue for that and link it here?

Reply:
Good point. However, once we move to mio-based asynchrnous sockets (my next PR) all sockets will become asynchronous, so maybe no need for the issue.

I will work on a standalone PR right after this one to switch the underlying sockets in `syscomm` to use `mio`.